File: leetcode/p00890/p00890.go
package p00890

// Time Complexity: O(nk)
// Space Complexity: O(2k+n)
func findAndReplacePattern(words []string, pattern string) []string {
	if len(pattern) == 1 {
		return words
	}

	result := []string{}
	for _, word := range words {
		if match(pattern, word) {
			result = append(result, word)
		}
	}

	return result
}

func match(pattern, word string) bool {
	mapping := map[byte]byte{}
	visited := map[byte]bool{}

	for i := 0; i < len(pattern); i++ {
		if !visited[pattern[i]] {
			if _, exists := mapping[word[i]]; exists {
				return false
			}
			visited[pattern[i]] = true
			mapping[word[i]] = pattern[i]

		} else {
			if mapping[word[i]] != pattern[i] {
				return false
			}
		}
	}

	return true
}

File: leetcode/p00890/p00890.py
from typing import List

class Solution:
    # Time Complexity: O(nk), wherer k is the length of the word
    # Space Complexity: O(n+2k), where k is the length of the word
    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        def match(word) -> bool:
            mapping = {}
            for i, j in zip(pattern, word):
                if mapping.setdefault(i, j) != j:
                    return False
            return len(set(mapping.values())) == len(mapping.values())

        return filter(match, words)

if __name__ == '__main__':
    s = Solution()

    assert s.findAndReplacePattern(["abc", "deq", "mee", "aqq", "dkd", "ccc"], "abb") == ["mee", "aqq"]
    assert s.findAndReplacePattern(["a", "b", "c"], "a") == ["a", "b", "c"]